lib/portage/package/ebuild/fetch.py

```diff
@@ -253,6 +260,130 @@ def _check_distfile(filename, digests, eout, show_errors=1, hash_filter=None):
 	'Y' : 80,
 }
 
+
+class FlatLayout(object):
+	def get_path(self, filename):
+		return filename
+
+
+class FilenameHashLayout(object):
+	def __init__(self, algo, cutoffs):
+		self.algo = algo
+		self.cutoffs = [int(x) for x in cutoffs.split(':')]
+
+	def get_path(self, filename):
+		fnhash = checksum_str(filename.encode('utf8'), self.algo)
+		ret = ''
+		for c in self.cutoffs:
+			assert c % 4 == 0
+			c = c // 4
+			ret += fnhash[:c] + '/'
+			fnhash = fnhash[c:]
+		return ret + filename
+
+
+class MirrorLayoutConfig(object):
+	"""
+	Class to read layout.conf from a mirror.
+	"""
+
+	def __init__(self):
+		self.structure = ()
+
+	def read_from_file(self, f):
+		cp = SafeConfigParser()
+		read_configs(cp, [f])
```

Comment:
We need to handle `configparser.ParsingError` somehow.

Reply:
Hmm, I've did a quick test earlier and presumed it won't fail. But now I managed to mess up the config file enough for it to fail ;-). So I'll handle it.

Reply:
I went for letting the caller care for `ConfigParserException`. In emirrordist, we will want to know when mirror config is broken.